rafthttp/http.go

```diff
@@ -122,6 +134,93 @@ func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusNoContent)
 }
 
+type snapshotHandler struct {
+	r         Raft
+	snapSaver SnapshotSaver
+	cid       types.ID
+}
+
+// ServeHTTP serves HTTP request to receive and process snapshot message.
+//
+// If request sender dies without closing underlying TCP connection,
+// the handler will keep waiting for the request body until TCP keepalive
+// finds out that the connection is broken after several minutes.
+// This is acceptable because
+// 1. following snapshot messages could still be received and processed
+// by ServeHTTP for other TCP connections.
+// 2. this case should happen rarely, so no further optimization is done.
+func (h *snapshotHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		w.Header().Set("Allow", "POST")
+		http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if err := checkVersionCompability(r.Header.Get("X-Server-From"), serverVersion(r.Header), minClusterVersion(r.Header)); err != nil {
+		plog.Errorf("request received was ignored (%v)", err)
+		http.Error(w, errIncompatibleVersion.Error(), http.StatusPreconditionFailed)
+		return
+	}
+
+	wcid := h.cid.String()
+	w.Header().Set("X-Etcd-Cluster-ID", wcid)
+
+	gcid := r.Header.Get("X-Etcd-Cluster-ID")
+	if gcid != wcid {
+		plog.Errorf("request received was ignored (cluster ID mismatch got %s want %s)", gcid, wcid)
+		http.Error(w, errClusterIDMismatch.Error(), http.StatusPreconditionFailed)
+		return
+	}
+
+	var l uint64
+	// read the length of raft message
+	if err := binary.Read(r.Body, binary.BigEndian, &l); err != nil {
```

Comment:
The format is `len | binary_proto_msg`. I check the encoder and decoder again, and it fits the format here. I resisted to reuse them before because encodoer/decoder is originally designed for stream, but it is no harm to use them when reading over this again.